packages/@lwc/engine-core/src/framework/hooks.ts

```diff
@@ -192,12 +216,13 @@ export function createViewModelHook(elm: HTMLElement, vnode: VCustomElement) {
         setElementShadowToken(elm, shadowAttribute);
     }
     const def = getComponentInternalDef(ctor);
-    createVM(elm, def, {
+    const vm = createVM(elm, def, {
         mode,
         owner,
         tagName: sel,
         renderer: owner.renderer,
     });
+    assertNotSyntheticComposedWithinNative(vm);
```

Comment:
I would recommend moving this logic in the `initializeMode`. The `createVM` call should fail if the native/synthetic shadow DOM composition is incorrect.